File: Java/46.permutations.java
/*
 * @lc app=leetcode id=46 lang=java
 *
 * [46] Permutations
 */

// @lc code=start
import java.util.*;
class Solution {
    /*
    permutation需要一个visited数组记录已访问过的数，dfs进入下一层的时候不需要记录当前pos
    time: O(n!)
    space: O(n)
    */
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();
        boolean[] visited = new boolean[nums.length];
        dfs(nums, visited, res, new ArrayList<>());
        return res;
    }

    private void dfs(int[] nums, boolean[] visited, List<List<Integer>> res, List<Integer> cur) {
        if (cur.size() == nums.length) {
            res.add(new ArrayList<>(cur));
            return;
        }
        for (int i = 0; i < nums.length; ++i) {
            if (!visited[i]) {
                cur.add(nums[i]);
                visited[i] = true;
                dfs(nums, visited, res, cur);
                cur.remove(cur.size() - 1);
                visited[i] = false;
            }
        }
    }
}
// @lc code=end

File: Cpp/46.permutations.cpp
/*
 * @lc app=leetcode id=46 lang=cpp
 *
 * [46] Permutations
 */

// @lc code=start
class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> res;
        vector<bool> visited(nums.size(), false);
        vector<int> cur;
        dfs(nums, res, visited, cur, 0);
        return res;
    }

private:
    void dfs(vector<int>& nums, vector<vector<int>>& res, vector<bool>& visited, vector<int>&cur, int pos) {
        if (pos == nums.size()) {
            res.push_back(cur);
            return;
        }
        for (int i = 0; i < nums.size(); ++i) {
            if (!visited[i]) {
                visited[i] = true;
                cur.push_back(nums[i]);
                dfs(nums, res, visited, cur, pos + 1);
                cur.pop_back();
                visited[i] = false;
            }
        }
    }
};
// @lc code=end

